File: core/src/main/java/dev/vibris/core/ArtifactJobTransaction.java
package dev.vibris.core;

import dev.vibris.api.ArtifactSink;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.channels.Channels;
import java.nio.channels.FileChannel;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import static java.nio.file.LinkOption.NOFOLLOW_LINKS;
import static java.nio.file.StandardOpenOption.CREATE_NEW;
import static java.nio.file.StandardOpenOption.WRITE;

class ArtifactJobTransaction implements ArtifactSink, AutoCloseable {
    private static final String MANIFEST = "manifest.json";

    private final ArtifactManager manager;
    private final Path temporary;
    private final Path directory;
    private final OwnedPathIdentity workspaceIdentity;
    private final OwnedPathIdentity temporaryIdentity;
    private final Map<String, ArtifactRecord> artifacts = new LinkedHashMap<>();
    private final Set<String> artifactNames = new HashSet<>();
    private final Set<ArtifactOutputStream> openStreams = new HashSet<>();
    private State state = State.ACTIVE;
    private long artifactBytes;
    private boolean writeFailed;

    ArtifactJobTransaction(ArtifactManager manager, Path temporary, Path directory,
                           OwnedPathIdentity workspaceIdentity) throws IOException {
        this.manager = manager;
        this.temporary = temporary;
        this.directory = directory;
        this.workspaceIdentity = workspaceIdentity;
        temporaryIdentity = OwnedPathIdentity.captureDirectory(temporary);
    }

    final Path directory() {
        return directory;
    }

    final OwnedPathIdentity workspaceIdentity() {
        return workspaceIdentity;
    }

    @Override
    public synchronized OutputStream open(String artifactName) throws IOException {
        requireActive();
        Path path = artifactPath(artifactName);
        String canonicalName = canonical(artifactName);
        if (!artifactNames.add(canonicalName)) {
            throw new IOException("Artifact name is already open or written.");
        }
        try {
            ArtifactOutputStream output = new ArtifactOutputStream(this, artifactName, path);
            openStreams.add(output);
            return output;
        } catch (IOException | RuntimeException exception) {
            artifactNames.remove(canonicalName);
            throw exception;
        }
    }

    public synchronized ArtifactManager.CommittedJob commit() throws IOException {
        return commit(null);
    }

    public synchronized ArtifactManager.CommittedJob commit(Set<String> expectedArtifacts) throws IOException {
        requireActive();
        if (!openStreams.isEmpty()) throw new IOException("Artifact streams must be closed before commit.");
        if (writeFailed) throw new IOException("Artifact output did not flush successfully.");
        if (expectedArtifacts != null) requireExpectedArtifacts(expectedArtifacts);
        verifyTemporaryAndArtifacts();
        state = State.FINALIZING;
        Map<String, Long> artifactSizes = artifactSizes();
        byte[] manifestBytes = ArtifactManifest.encode(artifactSizes);
        try {
            manager.reserve(this, Math.addExact(artifactBytes, manifestBytes.length));
            Path manifest = temporary.resolve(MANIFEST);
            ArtifactFiles.RegularFileIdentity manifestIdentity;
            try (FileChannel channel = FileChannel.open(manifest, CREATE_NEW, WRITE, NOFOLLOW_LINKS);
                 OutputStream output = Channels.newOutputStream(channel)) {
                manifestIdentity = ArtifactFiles.captureRegularFile(manifest);
                output.write(manifestBytes);
                output.flush();
                channel.force(true);
            }
            ArtifactFiles.verifyIdentity(manifest, manifestIdentity);
            manifestIdentity = ArtifactFiles.captureRegularFile(manifest);
            ArtifactFiles.verifiedSize(manifest, manifestIdentity, manifestBytes.length);
            long completedAt = Files.getLastModifiedTime(manifest).toMillis();
            verifyTemporaryAndArtifacts();
            ArtifactFiles.verifiedSize(manifest, manifestIdentity, manifestBytes.length);
            manager.verifyStorageIdentity(temporary.getParent(), workspaceIdentity);
            try {
                Files.move(temporary, directory, StandardCopyOption.ATOMIC_MOVE);
            } catch (AtomicMoveNotSupportedException exception) {
                throw new IOException("Artifact filesystem does not support atomic job finalize.", exception);
            }
            long totalBytes = Math.addExact(artifactBytes, manifestBytes.length);
            state = State.COMMITTED;
            manager.complete(this, totalBytes, completedAt);
            Map<String, Path> finalArtifacts = new LinkedHashMap<>();
            artifacts.keySet().forEach(name -> finalArtifacts.put(name, directory.resolve(name)));
            Map<String, Long> fileByteSizes = new LinkedHashMap<>(artifactSizes);
            fileByteSizes.put(MANIFEST, (long) manifestBytes.length);
            return new ArtifactManager.CommittedJob(
                directory, directory.resolve(MANIFEST), finalArtifacts, fileByteSizes, totalBytes);
        } catch (IOException | RuntimeException exception) {
            try {
                abortInternal();
            } catch (IOException cleanupFailure) {
                exception.addSuppressed(cleanupFailure);
            }
            throw exception;
        }
    }

    @Override
    public synchronized void close() throws IOException {
        if (state == State.COMMITTED || state == State.ABORTED) return;
        IOException failure = null;
        for (ArtifactOutputStream stream : List.copyOf(openStreams)) {
            try {
                stream.close();
            } catch (IOException exception) {
                if (failure == null) failure = exception;
                else failure.addSuppressed(exception);
            }
        }
        try {
            abortInternal();
        } catch (IOException exception) {
            if (failure == null) failure = exception;
            else failure.addSuppressed(exception);
        }
        if (failure != null) throw failure;
    }

    private Path artifactPath(String artifactName) throws IOException {
        if (artifactName == null || artifactName.isBlank() || artifactName.equals(".") ||
            artifactName.equals("..") || canonical(artifactName).equals(MANIFEST)) {
            throw new IOException("Artifact name must be one safe file name.");
        }
        try {
            if (!Path.of(artifactName).getFileName().toString().equals(artifactName)) {
                throw new IOException("Artifact name must be one safe file name.");
            }
        } catch (InvalidPathException exception) {
            throw new IOException("Artifact name is invalid.", exception);
        }
        Path path = temporary.resolve(artifactName).normalize();
        if (!path.getParent().equals(temporary)) throw new IOException("Artifact path escapes its job directory.");
        return path;
    }

    synchronized void write(ArtifactOutputStream stream, byte[] bytes, int offset, int length)
        throws IOException {
        requireActive();
        manager.reserve(this, Math.addExact(artifactBytes, length));
        try {
            stream.writeDirect(bytes, offset, length);
        } catch (IOException exception) {
            writeFailed = true;
            throw exception;
        }
        artifactBytes = Math.addExact(artifactBytes, length);
        stream.addBytes(length);
    }

    synchronized void closed(ArtifactOutputStream stream, boolean failed) {
        openStreams.remove(stream);
        artifacts.put(stream.name(), new ArtifactRecord(stream.path(), stream.bytes(), stream.identity()));
        if (failed) writeFailed = true;
    }

    private void requireActive() throws IOException {
        if (state != State.ACTIVE) throw new IOException("Artifact job is not active.");
    }

    private void abortInternal() throws IOException {
        state = State.ABORTED;
        manager.abort(this);
        manager.verifyStorageIdentity(temporary.getParent(), workspaceIdentity);
        if (!temporaryIdentity.matchesDirectory(temporary)) {
            throw new IOException("Artifact job directory changed before cleanup.");
        }
        ArtifactFiles.deleteTree(temporary, workspaceIdentity);
    }

    private void verifyTemporaryAndArtifacts() throws IOException {
        manager.verifyStorageIdentity(temporary.getParent(), workspaceIdentity);
        if (!temporaryIdentity.matchesDirectory(temporary)) {
            throw new IOException("Artifact job directory changed before finalize.");
        }
        for (ArtifactRecord artifact : artifacts.values()) {
            ArtifactFiles.verifiedSize(artifact.path, artifact.identity, artifact.bytes);
        }
    }

    private void requireExpectedArtifacts(Set<String> expectedArtifacts) throws IOException {
        Set<String> expected = new HashSet<>();
        for (String name : Objects.requireNonNull(expectedArtifacts, "expectedArtifacts")) {
            artifactPath(name);
            if (!expected.add(canonical(name))) throw new IOException("Expected artifact names are repeated.");
        }
        if (!expected.equals(artifactNames)) throw new IOException("Runtime did not write the expected artifacts.");
    }

    private Map<String, Long> artifactSizes() {
        Map<String, Long> sizes = new LinkedHashMap<>();
        artifacts.forEach((name, artifact) -> sizes.put(name, artifact.bytes));
        return sizes;
    }

    private static String canonical(String name) {
        return name.toLowerCase(Locale.ROOT);
    }

    private record ArtifactRecord(Path path, long bytes, ArtifactFiles.RegularFileIdentity identity) {
    }

    private enum State { ACTIVE, FINALIZING, COMMITTED, ABORTED }
}

File: core/src/main/kotlin/dev/vibris/core/ArtifactJobTransaction.kt
package dev.vibris.core

import dev.vibris.api.ArtifactSink
import java.io.IOException
import java.io.OutputStream
import java.nio.channels.Channels
import java.nio.channels.FileChannel
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.InvalidPathException
import java.nio.file.LinkOption.NOFOLLOW_LINKS
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption.CREATE_NEW
import java.nio.file.StandardOpenOption.WRITE
import java.util.LinkedHashMap
import java.util.Locale

internal open class ArtifactJobTransaction internal constructor(
    private val manager: ArtifactManager,
    private val temporary: Path,
    private val directoryValue: Path,
    private val workspaceIdentityValue: OwnedPathIdentity,
    private val workspaceId: String,
    private val jobId: String,
    private val requestId: String,
    private val recipe: String,
    private val createdAtUnixMs: Long,
    private val expiresAtUnixMs: Long,
) : ArtifactSink, AutoCloseable {
    private val temporaryIdentity = OwnedPathIdentity.captureDirectory(temporary)
    private val artifacts = LinkedHashMap<String, ArtifactRecord>()
    private val artifactNames = HashSet<String>()
    private val openStreams = HashSet<ArtifactOutputStream>()
    private var state = State.ACTIVE
    private var artifactBytes = 0L
    private var writeFailed = false

    internal fun directory(): Path = directoryValue

    internal fun workspaceIdentity(): OwnedPathIdentity = workspaceIdentityValue

    @Synchronized
    @Throws(IOException::class)
    override fun open(artifactName: String): OutputStream {
        requireActive()
        val path = artifactPath(artifactName)
        val canonicalName = canonical(artifactName)
        if (!artifactNames.add(canonicalName)) {
            throw IOException("Artifact name is already open or written.")
        }
        try {
            val output = ArtifactOutputStream(this, artifactName, path)
            openStreams.add(output)
            return output
        } catch (exception: IOException) {
            artifactNames.remove(canonicalName)
            throw exception
        } catch (exception: RuntimeException) {
            artifactNames.remove(canonicalName)
            throw exception
        }
    }

    @Synchronized
    fun checkpoint(): Checkpoint {
        requireActive()
        return Checkpoint(java.util.Set.copyOf(artifactNames), artifactBytes, writeFailed)
    }

    @Synchronized
    @Throws(IOException::class)
    fun rollback(checkpoint: Checkpoint) {
        requireActive()
        var failure: IOException? = null
        for (stream in java.util.List.copyOf(openStreams)) {
            if (canonical(stream.name()) in checkpoint.artifactNames) continue
            try {
                stream.close()
            } catch (exception: IOException) {
                failure = appendFailure(failure, exception)
            }
        }
        val created = artifacts.entries.filter { canonical(it.key) !in checkpoint.artifactNames }
        for ((name, artifact) in created) {
            try {
                ArtifactFiles.verifiedSize(artifact.path, artifact.identity, artifact.bytes)
                Files.delete(artifact.path)
                artifacts.remove(name)
                artifactNames.remove(canonical(name))
            } catch (exception: IOException) {
                failure = appendFailure(failure, exception)
            }
        }
        if (failure != null) throw failure
        artifactBytes = checkpoint.artifactBytes
        writeFailed = checkpoint.writeFailed
    }

    @Synchronized
    @Throws(IOException::class)
    fun commit(specifications: Map<String, ArtifactManifest.FileSpec>): ArtifactManager.CommittedJob {
        requireActive()
        if (openStreams.isNotEmpty()) {
            throw IOException("Artifact streams must be closed before commit.")
        }
        if (writeFailed) {
            throw IOException("Artifact output did not flush successfully.")
        }
        requireSpecifications(specifications)
        verifyTemporaryAndArtifacts()
        state = State.FINALIZING
        val artifactSizes = artifactSizes()
        val completedAt = manager.now()
        val entries = artifacts.map { (name, artifact) ->
            ArtifactManifest.FileEntry(
                manager.artifactId(workspaceId, jobId, requestId, name),
                name,
                specifications.getValue(name),
                artifact.bytes,
                ArtifactManifest.sha256(artifact.path),
                createdAtUnixMs,
                expiresAtUnixMs,
            )
        }
        val encoded = ArtifactManifest.encode(
            ArtifactManifest.Document(
                manager.manifestId(workspaceId, jobId, requestId),
                workspaceId,
                jobId,
                requestId,
                recipe,
                createdAtUnixMs,
                completedAt,
                expiresAtUnixMs,
                0,
                entries,
            ),
        )
        val manifestBytes = encoded.bytes
        try {
            manager.reserve(this, encoded.document.totalBytes)
            val manifest = temporary.resolve(MANIFEST)
            lateinit var manifestIdentity: ArtifactFiles.RegularFileIdentity
            FileChannel.open(manifest, CREATE_NEW, WRITE, NOFOLLOW_LINKS).use { channel ->
                Channels.newOutputStream(channel).use { output ->
                    manifestIdentity = ArtifactFiles.captureRegularFile(manifest)
                    output.write(manifestBytes)
                    output.flush()
                    channel.force(true)
                }
            }
            ArtifactFiles.verifyIdentity(manifest, manifestIdentity)
            manifestIdentity = ArtifactFiles.captureRegularFile(manifest)
            ArtifactFiles.verifiedSize(manifest, manifestIdentity, manifestBytes.size.toLong())
            verifyTemporaryAndArtifacts()
            ArtifactFiles.verifiedSize(manifest, manifestIdentity, manifestBytes.size.toLong())
            manager.verifyStorageIdentity(temporary.parent, workspaceIdentityValue)
            try {
                Files.move(temporary, directoryValue, StandardCopyOption.ATOMIC_MOVE)
            } catch (exception: AtomicMoveNotSupportedException) {
                throw IOException("Artifact filesystem does not support atomic job finalize.", exception)
            }
            state = State.COMMITTED
            manager.complete(this, encoded)
            val finalArtifacts = LinkedHashMap<String, Path>()
            artifacts.keys.forEach { name -> finalArtifacts[name] = directoryValue.resolve(name) }
            val fileByteSizes = LinkedHashMap(artifactSizes)
            fileByteSizes[MANIFEST] = manifestBytes.size.toLong()
            val metadata = LinkedHashMap<String, dev.vibris.protocol.v2.ArtifactMetadata>()
            entries.forEach { entry ->
                metadata[entry.relativePath] = entry.metadata(jobId, requestId, directoryValue.resolve(entry.relativePath))
            }
            return ArtifactManager.CommittedJob(
                directoryValue,
                directoryValue.resolve(MANIFEST),
                finalArtifacts,
                fileByteSizes,
                metadata,
                encoded.document.manifestId,
                encoded.sha256,
                expiresAtUnixMs,
                encoded.document.totalBytes,
            )
        } catch (exception: IOException) {
            abortAfterFailure(exception)
        } catch (exception: RuntimeException) {
            abortAfterFailure(exception)
        }
    }

    @Synchronized
    @Throws(IOException::class)
    internal fun readableArtifact(artifactName: String?): Path {
        requireActive()
        val artifact = artifacts[artifactName]
            ?: throw IOException("Artifact is not available for comparison.")
        ArtifactFiles.verifiedSize(artifact.path, artifact.identity, artifact.bytes)
        return artifact.path
    }

    @Synchronized
    @Throws(IOException::class)
    override fun close() {
        if (state == State.COMMITTED || state == State.ABORTED) {
            return
        }
        var failure: IOException? = null
        for (stream in java.util.List.copyOf(openStreams)) {
            try {
                stream.close()
            } catch (exception: IOException) {
                if (failure == null) {
                    failure = exception
                } else {
                    failure?.addSuppressed(exception)
                }
            }
        }
        try {
            abortInternal()
        } catch (exception: IOException) {
            if (failure == null) {
                failure = exception
            } else {
                failure?.addSuppressed(exception)
            }
        }
        failure?.let { throw it }
    }

    @Synchronized
    @Throws(IOException::class)
    internal fun write(stream: ArtifactOutputStream, bytes: ByteArray, offset: Int, length: Int) {
        requireActive()
        manager.reserve(this, Math.addExact(artifactBytes, length.toLong()))
        try {
            stream.writeDirect(bytes, offset, length)
        } catch (exception: IOException) {
            writeFailed = true
            throw exception
        }
        artifactBytes = Math.addExact(artifactBytes, length.toLong())
        stream.addBytes(length)
    }

    @Synchronized
    internal fun closed(stream: ArtifactOutputStream, failed: Boolean) {
        openStreams.remove(stream)
        artifacts[stream.name()] = ArtifactRecord(stream.path(), stream.bytes(), stream.identity())
        if (failed) {
            writeFailed = true
        }
    }

    @Throws(IOException::class)
    private fun artifactPath(artifactName: String?): Path {
        if (
            artifactName == null ||
            artifactName.isBlank() ||
            artifactName == "." ||
            artifactName == ".." ||
            canonical(artifactName) == MANIFEST
        ) {
            throw IOException("Artifact name must be one safe file name.")
        }
        try {
            if (Path.of(artifactName).fileName.toString() != artifactName) {
                throw IOException("Artifact name must be one safe file name.")
            }
        } catch (exception: InvalidPathException) {
            throw IOException("Artifact name is invalid.", exception)
        }
        val path = temporary.resolve(artifactName).normalize()
        if (path.parent != temporary) {
            throw IOException("Artifact path escapes its job directory.")
        }
        return path
    }

    @Throws(IOException::class)
    private fun requireActive() {
        if (state != State.ACTIVE) {
            throw IOException("Artifact job is not active.")
        }
    }

    @Throws(IOException::class)
    private fun abortInternal() {
        state = State.ABORTED
        manager.abort(this)
        manager.verifyStorageIdentity(temporary.parent, workspaceIdentityValue)
        if (!temporaryIdentity.matchesDirectory(temporary)) {
            throw IOException("Artifact job directory changed before cleanup.")
        }
        ArtifactFiles.deleteTree(temporary, workspaceIdentityValue)
    }

    @Throws(IOException::class)
    private fun verifyTemporaryAndArtifacts() {
        manager.verifyStorageIdentity(temporary.parent, workspaceIdentityValue)
        if (!temporaryIdentity.matchesDirectory(temporary)) {
            throw IOException("Artifact job directory changed before finalize.")
        }
        for (artifact in artifacts.values) {
            ArtifactFiles.verifiedSize(artifact.path, artifact.identity, artifact.bytes)
        }
    }

    @Throws(IOException::class)
    private fun requireSpecifications(specifications: Map<String, ArtifactManifest.FileSpec>) {
        val expected = HashSet<String>()
        for ((name, specification) in specifications) {
            artifactPath(name)
            if (!expected.add(canonical(name))) throw IOException("Artifact specifications are repeated.")
            if (specification.kind.name.endsWith("UNSPECIFIED") ||
                specification.format.name.endsWith("UNSPECIFIED") ||
                specification.role.name.endsWith("UNSPECIFIED") ||
                specification.mediaType.isBlank()
            ) throw IOException("Artifact specification is incomplete.")
        }
        if (expected != artifactNames) throw IOException("Runtime did not write the specified artifacts.")
    }

    private fun artifactSizes(): Map<String, Long> {
        val sizes = LinkedHashMap<String, Long>()
        artifacts.forEach { (name, artifact) -> sizes[name] = artifact.bytes }
        return sizes
    }

    private fun abortAfterFailure(exception: Exception): Nothing {
        try {
            abortInternal()
        } catch (cleanupFailure: IOException) {
            exception.addSuppressed(cleanupFailure)
        }
        throw exception
    }

    private data class ArtifactRecord(
        val path: Path,
        val bytes: Long,
        val identity: ArtifactFiles.RegularFileIdentity,
    )

    data class Checkpoint internal constructor(
        internal val artifactNames: Set<String>,
        internal val artifactBytes: Long,
        internal val writeFailed: Boolean,
    )

    private enum class State {
        ACTIVE,
        FINALIZING,
        COMMITTED,
        ABORTED,
    }

    private companion object {
        const val MANIFEST = "manifest.json"

        fun canonical(name: String): String = name.lowercase(Locale.ROOT)

        private fun appendFailure(original: IOException?, additional: IOException): IOException {
            if (original == null) return additional
            original.addSuppressed(additional)
            return original
        }
    }
}
